apps/minifront/src/components/staking/account/use-staking-tokens-and-filter.ts

```diff
@@ -0,0 +1,91 @@
+import { useMemo } from 'react';
+import {
+  BalancesByAccount,
+  balancesByAccountSelector,
+  useBalancesResponses,
+  useStakingTokenMetadata,
+} from '../../../state/shared';
+import {
+  Metadata,
+  ValueView,
+} from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/asset/v1/asset_pb';
+import { getDisplayDenomFromView } from '@penumbra-zone/getters/value-view';
+import { assetPatterns } from '@penumbra-zone/types/assets';
+
+/**
+ * Function to use with `reduce()` over an array of `BalancesByAccount` objects.
+ * Returns a map of accounts to `ValueView`s of the staking token.
+ */
+const toStakingTokensByAccount = (
+  stakingTokensByAccount: Map<number, ValueView | undefined>,
+  curr: BalancesByAccount,
+  stakingTokenMetadata: Metadata,
+) => {
+  const stakingTokenBalance = curr.balances.find(
+    ({ balanceView }) => getDisplayDenomFromView(balanceView) === stakingTokenMetadata.display,
+  );
+
+  if (stakingTokenBalance?.balanceView) {
+    stakingTokensByAccount.set(curr.account, stakingTokenBalance.balanceView);
+  }
+
+  return stakingTokensByAccount;
+};
+
+/**
+ * Function to use with `reduce()` over an array of `BalancesByAccount` objects.
+ * Reduces to an array of accounts that have relevant balances for staking.
+ */
+const toAccountSwitcherFilter = (
+  accountSwitcherFilter: number[],
+  curr: BalancesByAccount,
+  stakingTokenMetadata: Metadata,
+) => {
+  const isRelevantAccount = curr.balances.some(({ balanceView }) => {
+    const displayDenom = getDisplayDenomFromView(balanceView);
+
+    return (
+      assetPatterns.delegationToken.matches(displayDenom) ||
+      assetPatterns.unbondingToken.matches(displayDenom) ||
+      displayDenom === stakingTokenMetadata.display
+    );
+  });
+
+  if (isRelevantAccount) return [...accountSwitcherFilter, curr.account];
+  return accountSwitcherFilter;
+};
+
+export const useStakingTokensAndFilter = (
```

Comment:
This hook is basically doing what the `stakingTokensAndFilter` ZQuery object used to be doing. But since that object was just using balances and the staking token metadata, which are now both in the `shared` slice, this hook uses the `shared` slice as its source of data, and massages the data into the shape expected by the consumers of the hook.